semeio/workflows/ahm_analysis/ahmanalysis.py

```diff
@@ -0,0 +1,456 @@
+
+    """To get the inactive vs total observation info."""
+    df_active = df_update_log[(df_update_log["status"] == "Active")]
+    return str(len(df_active)) + " active/" + str(len(df_update_log.index))
+
+
+def create_path(runpathdf, target_name):
+    """to create path for output data"""
+    return tempfile.mkdtemp()
+
+
+def list_observations_misfit(obs_keys, df_update_log, misfit_df):
+    """To get the misfit for total observations (active/inactive)."""
+    misfit_values = []
+    total_obs_nr = len(df_update_log[df_update_log.status.isin(["Active", "Inactive"])])
+    for key in obs_keys:
+        misfit_name = "MISFIT:" + key
+        misfit_values.append(misfit_df[misfit_name].mean())
+    return np.sum(misfit_values) / total_obs_nr
+
+
+def get_field_grid_char(input_grid):
+    """Get field grid characteristics/coordinates"""
+    input_grid = Path(input_grid).with_suffix("")
+    try:
+        mygrid = Grid(input_grid, fformat="eclipserun")
+        dataframeg = mygrid.dataframe(activeonly=False)
+        return dataframeg
+    except OSError as err:
+        raise OSError("A grid with .EGRID format is expected.") from err
+
+
+def calc_delta_grid(
+    all_input_post, all_input_prior, mygrid_ok, caseobs, mygrid_ok_short
+):
+    """calculate mean delta of field grid data"""
+    delta_post_prior = np.subtract(all_input_prior, all_input_post)
+    delta_post_prior = np.absolute(delta_post_prior)
+    # Can we avoid changing in place here?
+    mygrid_ok["Mean_delta"] = np.mean(delta_post_prior, axis=0)
+    df_mean_delta = mygrid_ok.groupby(["IX", "JY"])[["Mean_delta"]].mean().reset_index()
+    mygrid_ok_short["Mean_D_" + caseobs] = df_mean_delta["Mean_delta"]
+    return mygrid_ok_short
+
+
+def _get_parameter_keys(genkw_parameters, prior_keys):
+    """excluding duplicate transformed parameter keys"""
+    transformed_param_keys = []
+    for parameter in genkw_parameters:
+        for key in prior_keys:
+            if "_" + parameter + ":" in key:
+                transformed_param_keys.append(key)
+    return set(prior_keys).difference(transformed_param_keys)
+
+
+def get_updated_parameters(prior_data, parameters):
+    """make list of updated parameters
+    (excluding duplicate transformed parameters)
+    """
+    parameter_keys = _get_parameter_keys(parameters, prior_data.keys())
+    # remove parameters with constant prior distribution
+    p_keysf = []
+    for dkey in parameter_keys:
+        if prior_data[dkey].ndim > 1:
+            warnings.warn("WARNING: Parameter " + dkey + " defined several times.")
+            flatten_arr = np.ravel(prior_data[dkey])
+            result = np.all(prior_data[dkey] == flatten_arr[0])
+            if not result:
+                p_keysf.append(dkey)
+        else:
+            if not all(x == prior_data[dkey][0] for x in prior_data[dkey]):
+                p_keysf.append(dkey)
+    return p_keysf
+
+
+def calc_ks(ks_matrix, dkeys, prior_data, target_data, key):
```

Comment:
We should either modify or return a new instance, not both.

Reply:
option without comprehension is now implemented